Refactoring/unit testing for x-powered-by scanner, cookie-same-site scanner, and gitignore in alpha branch

vahidR opened the pull request:
The refactoring consists of
1. extending the PassiveScannerTest class to avoid the duplications
2. Raising only one alert if XPoweredByHeaderInfoLeakScanner finds
multiple x-powered-by headers
3. In case of 2, it populates the <Evidence field> with the first header
found and the <Other field> with the rest of found values; It worthwhile
to mention that
4. Updating gitignore on alpha branch, to ignore IntelliJ specific files

psiinon commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,111 @@
+package org.zaproxy.zap.extension.pscanrulesAlpha;
```
Should include the standard header

vahidR replied:
Oh, I missed that. 
Fixed..

psiinon commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,111 @@
+ */
+
+@RunWith(MockitoJUnitRunner.class)
+public class XPoweredByHeaderInfoLeakScannerUnitTest {
```
I've updated my PR #473 based on comments to introduce an abstract class, would be worth doing the same here

vahidR replied:
@psiinon I couldn't find any discussion about abstract class. 
However, I saw the one about helper method that makes sense to some extent. 
Would you please explain about abstract class?

psiinon replied:
Heres the abstract class I created: https://github.com/psiinon/zap-extensions/blob/master/test/org/zaproxy/zap/extension/pscanrules/PassiveScannerTest.java
And this is an example of it being extended: https://github.com/psiinon/zap-extensions/blob/master/test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java
As you'll see it removes a load of boilerplate code from the extending class :)

vahidR replied:
Nice.. I'll refactor / rebase my class then. 
Thanks :)

psiinon replied:
Thank you - more passive scan unit tests will be really useful :)
I think we should have a new policy that means we only promote passive scan tests if they have good unit tests (cc @thc202 @kingthorin )

vahidR replied:
No problem. I'll be glad to help and learn at the same time..

But some notes:
1. These classes are not merged into zaproxy/zap-extensions repo, and only exists on your repo. So merging it would be nice.
2. They only exist on the **master** branch. What's the policy about _alpha_ and _beta_ branches? Do you want me to add this class to the alpha branch? (in pscanrulesalpha package). 
I'm not quite sure about your policy on cherry picking commits or how sync these three branches are..

thc202 replied:
@psiinon requiring unit tests for promotion is a good plan.

@vahidR you can just copy the class, no issue there (it would have to be copied even if the pull request was already merged).
Once the branches are merged into one we can take a closer look at how to better handle the helper test classes.

psiinon commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,111 @@
+        Source source = new Source(response);
+
+        // When
+        rule.scanHttpResponseReceive(msg, rule.getPluginId(), source);
```
Thats a cnp bug from my code ;) rule.getPluginIn() isnt the right thing to supply here (even though its not checked) - I've changed to use -1

vahidR replied:
OK.. Fixed!

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,111 @@
+        msg.setResponseHeader(
+                "HTTP/1.1 200 OK\r\n" +
+                        "Server: Apache-Coyote/1.1\r\n" +
+                        "X-Powered-By: Servlet/3.0 " +
```
It's missing the header separator.

vahidR replied:
You are right... Fixed!

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,111 @@
+        // Then
+        assertThat(alertsRaised.size(), is(1));
+        assertThat(alertsRaised.get(0).getEvidence(), is("X-Powered-By"));
+        assertThat(alertsRaised.get(0).getCweId(), is(200));
```
Should also check the WASC ID?

vahidR replied:
Not really... removed!

psiinon commented:
I've also just updated #473 to reduce the unit test boiler plate code, so prob worth doing the same here

vahidR commented:
Alright.. I'll refactor my class to follow the same standard.

vahidR commented:
I refactored and rebased my code.
The commit includes:

1. Adding _PassiveScannerTest_ class into the _pscanrulesAlpha_ package.
2. Changing the _XPoweredByHeaderInfoLeakScannerUnitTest_ class to follow the new standard.
3. Adding IntelliJ specific files/folders to gitignore file in the alpha branch

psiinon commented:
Looks good

kingthorin commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+
+/**
+ * @author Vahid Rafiei (@vahid_r)
+ * Date 2016-07-06
```
This should be dropped if it isn't going to be maintained...

vahidR replied:
Although I will maintain my code, I will remove the date..

kingthorin replied:
:smile: yup, I just meant the one line...

vahidR replied:
aha.. OK :)

vahidR commented:
@psiinon @thc202 @kingthorin please don't merge it yet.. I'm refactoring the CookieSameSiteScannerUnitTest class to follow the new standard..

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+ * Date 2016-07-06
+ */
+
+@RunWith(MockitoJUnitRunner.class)
```
This annotation can be moved to base class.

vahidR replied:
Yes, will do that.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+                        "Content-Type: text/html;charset=ISO-8859-1\r\n" +
+                        "Content-Length: " + msg.getResponseBody().length() + "\r\n");
+        String response = msg.getResponseHeader().toString() + msg.getResponseBody().toString();
+        Source source = new Source(response);
```
The two statements can be replaced with a call to `createSource(msg)`;

vahidR replied:
I missed this method. Thanks

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+                        "Content-Type: text/html;charset=ISO-8859-1\r\n" +
+                        "Content-Length: " + msg.getResponseBody().length() + "\r\n");
+        String response = msg.getResponseHeader().toString() + msg.getResponseBody().toString();
+        Source source = new Source(response);
```
Same here.

vahidR replied:
yep

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
+
+        msg.setResponseBody("<html></html>");
```
IMO this could be dropped, the body is not important (i.e. not used) for this scanner.

vahidR replied:
Ok..

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,90 @@
+                "HTTP/1.1 200 OK\r\n" +
+                        "Server: Apache-Coyote/1.1\r\n" +
+                        "Content-Type: text/html;charset=ISO-8859-1\r\n" +
+                        "Content-Length: " + msg.getResponseBody().length() + "\r\n");
```
Per above comment the headers Content-Type and Content-Length could be dropped too.

vahidR replied:
I thought It's better to have have these, but we are interested in X-Powered-By. So I'll remove them..

thc202 commented:
Commit message needs some tweaks (remove some lines?).

It would be good to test the case when multiple X-Powered-By headers are present, as that leads to multiple alerts (but no longer with different evidence, I guess this would lead to some changes in the scanner, raise just one alert?).

vahidR commented:
> Commit message needs some tweaks (remove some lines?).
I Will do better in the coming commit.

> It would be good to test the case when multiple X-Powered-By headers are present....
Interesting case! It raises multiple alerts as you said. So do we prefer to raise only one alert in this situation?

vahidR commented:
I tried to apply your comments.

1. PassiveScannerTest has been added to the pscanrulesAlpha package. 
2. XPoweredByHeaderInfoLeakScannerUnitTest has been refactored.
3. XPoweredByHeaderInfoLeakScanner refactored to raise only one alert if it finds multiple x-powerd-by flag. 
4. CookieSameSiteScannerUnitTest refactored to follow the new standard.

vahidR commented:
@psiinon @thc202 @kingthorin Should I proceed with the last refactoring that I made in the  _XPoweredByHeaderInfoLeakScanner_ class? (i.e. only raising one alert if the scanner finds _multiple similar_ headers, but with the _same evidence_).
This will affect the future unit tests that I am going to write, as the refactoring logic will be applied in similar situations..

thc202 commented:
I'm not following, are you referring to the changes already applied to ` XPoweredByHeaderInfoLeakScanner` (i.e. keep what's already done) or if that should be applied to other scanners?

vahidR commented:
I mean bringing the refactoring that I've already done in `XPoweredByHeaderInfoLeakScanner`to other classes. 
For instance `XBackendServerInformationLeak` has a similar logic.

kingthorin commented:
I'm confused about the change to the scanner. If there are multiple headers, there are likely multiple things for the system owner/dev to fix. If we only alert on one, then they scan again and alert on another people will get annoyed. 

The previous scanner change (to highlight the header) was done because the evidence was sometimes both in XPB and Server and whichever appeared first would be highlighted which could lead to a misleading UX. It seems what actually need to be done is that the whole header + value is highlighted (evidence) not just one or the other.

vahidR commented:
The main idea behind the refactoring is to not raising alerts when there is the same evidence for every alert. 
Suggested by @thc202 I also think this is a right approach to raise only one. 
But I'm open to your suggestions anyway; just let me know what you think is better..

psiinon commented:
How about raising one alert with the first header as 'evidence' and then including all of the other headers (if any) in the 'other' field? Best of both worlds maybe?

vahidR commented:
By **other** field, do you mean the 5th argument of the Alert.setDetail() method ?

kingthorin commented:
> How about raising one alert with the first header as 'evidence' and then including all of the other headers (if any) in the 'other' field? Best of both worlds maybe?

I'm good with that idea.

> By other field, do you mean the 5th argument of the Alert.setDetail() method ? 

[Yup](https://github.com/vahidR/zap-extensions/blob/32b80e16d30d1c9eb805f796e2f5479cd3090380/src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java#L73).

vahidR commented:
Alright then... 
I will implement that and write the relevant unit tests for it.

psiinon commented:
FYI #473 has just been merged, so should be safe to grab the code from now ;)

vahidR commented:
Please review the changes and let me know if it needs more tweaks..

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,95 @@
+        // Then
+        assertThat(alertsRaised.size(), is(3));
+        assertThat(alertsRaised.get(0).getEvidence(), is("X-Powered-By"));
+        assertThat(alertsRaised.get(1).getOtherInfo(), is("X-Powered-By"));
```
It should raise just one alert, subsequent (or all?)  `X-Powered-By` headers (i.e. its values) would be added to the other info of the ("first") alert.

vahidR replied:
OK... I'll fix it.

kingthorin commented:
I think you misunderstood the earlier comments. It's going to be a single alert either way, if there are multiple XPB headers then the details of the first occurrence should be 'Evidence' the others should be included in the 'other info' field.

thc202 commented:
What should be in the evidence field, the whole header? Or, just the header name?

kingthorin commented:
Edited my comment to clarify...

vahidR commented:
Alright guys. I will fix it.

> What should be in the evidence field, the whole header? Or, just the header name?

I think the scenarios described by @kingthorin is the answer: 

> So 2 scenarios:
> 1. Single XPB, one alert with evidence as the whole XPB header (header field + value).
> 2. Multiple XPB, one alert with evidence as the whole first XPB header (header field + value), and other info contains the other XPB header(s) [all header fields + values].

kingthorin commented:
I believe the whole header is the best option.

Originally it had been the value, which caused problems and was why it was switched to only the header field. But to properly highlight the first header and then provide all the necessary details as other info I think it's best to use the whole header. i.e.: `X-Powered-By: PHP/5.5.9-1ubuntu4.14` or `X-Powered-By: ASP.NET`, etc.

thc202 commented:
OK, makes sense to me.

The summary/description of the pull request (and the commit) should be updated to mention the change of behaviour.

psiinon commented:
I'd much prefer just the XPH parts and not unrelated headers to be logged as evidence. They should also be an exact match, so that we can highlight exactly the right part of the response.

thc202 commented:
With the "whole header" I think it means the header name and value, not the whole response header, right?

psiinon commented:
Thats what I was hoping :)

vahidR commented:
Yes, I only filter the X-Powered-By values.

kingthorin commented:
> With the "whole header" I think it means the header name and value, not the whole response header, right?

Yup.

vahidR commented:
I did the changes; Please review it :)

kingthorin commented:
Looks good. 

I've updated the PR title and message to align with the commit details.

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -36,7 +37,7 @@
  * @author kingthorin+owaspzap@gmail.com
  */
-public class XPoweredByHeaderInfoLeakScanner extends PluginPassiveScanner{
+public class XPoweredByHeaderInfoLeakScanner extends PluginPassiveScanner {
```
The formatting change could be reverted. [1]

[1] https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
 	@Override
 	public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
 		long start = System.currentTimeMillis();
-	
```
Same here.

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
-				}
-			}
-		    if (logger.isDebugEnabled()) {
+        if (xpbOptions != null) {
```
It could use the same indentation char (tab) as surrounding code (same for other lines).

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+     * @param msg
+     * @param id
+     */
+	private void raiseAlert(HttpMessage msg, int id, List<String> optionValues) {
```
Consider renaming "optionValues" to "headerValues".
The JavaDoc could be finished (i.e. document all parameters),

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+     * @param id
+     */
+	private void raiseAlert(HttpMessage msg, int id, List<String> optionValues) {
+        String keyValueOptionForEvidenceField = "";
```
Consider renaming to just `alertEvidence`?

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+     */
+	private void raiseAlert(HttpMessage msg, int id, List<String> optionValues) {
+        String keyValueOptionForEvidenceField = "";
+        String keyValueOptionForOtherField = "";
```
Consider renaming to just `alertOtherInfo`?

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+        String keyValueOptionForEvidenceField = "";
+        String keyValueOptionForOtherField = "";
+        if (optionValues.size() == 1) {
+            keyValueOptionForEvidenceField = HEADER_NAME + ": " + optionValues.get(0);
```
- The statement is the same for both paths so it could be done outside the if-then statement (it could be done when the variable is declared instead of initialising to an empty string).
 - The evidence will not exactly match what's in the response header if the header name has a different case.

vahidR replied:
> The statement is the same for both paths so it could be done outside the if-then statement (it could be done when the variable is declared instead of initialising to an empty string).

Refactored..

> The evidence will not exactly match what's in the response header if the header name has a different case.

Do you mean if the response header is lower-case for instance? 
If so, you are right, as we cast everything to Upper-case within `getHeaders()` method. 
But the question is that do we really care? As we upper case the _key_ and will be interested in _values_ itself?

kingthorin replied:
I think he's concerned that if the evidence is not a specific match then the highlighting won't work correctly.

Instead of attempting to re-create the header(s) there must be a way to just use what was originally found...

vahidR replied:
I will think about it

kingthorin replied:
Thanks, I know that message was a bit obtuse. I don't have code or even a version of ZAP in front of me to look at. I'll see if I can come up with something more specific tonight or tomorrow...

vahidR replied:
We can do that, but it seems that we need to change the `HttpHeader` in zaproxy to make it so.
I suggest create a new issue and make it out of the scope of this PR, as the unit test relies on the current implementation..

thc202 replied:
Yes, that's about the matching/highlighting not working correctly when the header name has a different case.

It does not require changes in core, it can be obtained using a regex of the header name and the corresponding value. [1]
Of course, it should be possible to obtain the whole header (instead of just the value) using the `HttpHeader` class.

[1] https://github.com/zaproxy/zap-extensions/blob/7658674921d42850ff0bbe81a5acd7a0e1847f66/src/org/zaproxy/zap/extension/pscanrules/SetCookieUtils.java#L127-L145

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+            keyValueOptionForEvidenceField = HEADER_NAME + ": " + optionValues.get(0);
+        } else { // we have multiple values for the X-Powered-By key
+            keyValueOptionForEvidenceField = HEADER_NAME + ": " + optionValues.get(0);
+            StringBuilder sb = new StringBuilder();
```
It could have a message before the header enumeration explaining what the other info is/has (e.g. "The following X-Powered-By headers were also found: ..."?)

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +59,51 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+                sb.append(HEADER_NAME);
+                sb.append(": ");
+                sb.append(optionValues.get(i));
+                sb.append("; ");
```
Maybe newline instead? That improves the reading, as it's more clear when one header ends and starts the other.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScannerUnitTest.java:
```diff
@@ -21,67 +21,32 @@
 	}
 
-    @Test
+	@Test
```
This change could be reverted.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScannerUnitTest.java:
```diff
@@ -21,67 +21,32 @@
 		
 		HttpMessage msg = new HttpMessage();
 		msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
-		
```
This change could be reverted.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * @author Vahid Rafiei (@vahid_r)
+ */
+
```
This empty line could be removed.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @Test
+    public void shouldRaiseOnlyOneAlerWithOneEvidencetAndOtherInfoIfFindsMultipleXPoweredBy() throws Exception {
```
It's missing the "t" in "Alert".

thc202 commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 build/build/
 build/zap-exts/
+.idea/
+*.iml
```
These changes could be mentioned in the commit message.

thc202 commented:
Agreed, it would be great to also mention the change done to the other test class in commit/PR.

vahidR commented:
> Agreed, it would be great to also mention the change done to the other test class in commit/PR.

will do it..

vahidR commented:
I have done the refactoring, but I just saw that the rebase is not complete. 
I will fix it later, but please take a look at the favourite header extraction in `XPoweredByHeaderInfoLeakScanner` and tell me your idea..

vahidR commented:
I rebased it to a single commit ..

vahidR commented:
@thc202 Hi, any comment about my last changes? Please exclude the white spaces..

kingthorin commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+		for (String header : headerElements) {
+			if (matcher.find()) {
+				String match = matcher.group();
+				matchedHeaders.add(match);
```
Could just add `matcher.group()` instead of using an intermediary.

vahidR replied:
That is correct..

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
-		    	parent.raiseAlert(id, alert);
-				}
+
+		List<String> xpbHeaders = new ArrayList<>();
```
The declaration of the variable could be done inside the if statement (which does not require instantiating an unused ArrayList).

vahidR replied:
Ah, right.. refactored

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+        String alertEvidence = xpbHeaders.get(0);
+        String alertOtherInfo = "";
+        if (xpbHeaders.size() > 1) { // we have multiple X-Powered-By headers
+            StringBuilder sb = new StringBuilder();
```
It could set a higher capacity (the length of following string and the headers are more than the default).

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+        String alertOtherInfo = "";
+        if (xpbHeaders.size() > 1) { // we have multiple X-Powered-By headers
+            StringBuilder sb = new StringBuilder();
+			sb.append("The following X-Powered-By headers were also found:\r\n");
```
The string should be i18n.

vahidR replied:
Thanks.. learning something new everyday.
Refactored

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+			sb.append("The following X-Powered-By headers were also found:\r\n");
+            for (int i = 1; i < xpbHeaders.size(); i++) {
+                sb.append(xpbHeaders.get(i));
+                sb.append(";\r\n");
```
The semicolon could be removed?

vahidR replied:
removed..

thc202 commented:
The extraction of the header looks right to me.
Which white space chars are you referring to? Some were deleted, you would have to add them.

Added some comments, it would be good to have a test for different case of the header name.

vahidR commented:
> Which white space chars are you referring to? Some were deleted, you would have to add them.

Like the space that has injected before `sb.append("The following X-Powered-By header....");`, the line 115 of `XPoweredByHeaderInfoLeakScanner`. 
I can't see this indentation in my IDE..

thc202 commented:
Ah, that's because of tabs vs spaces, in the IDE it's probably the same length (i.e. 1 tab is 4 spaces).
Ideally we should just use one of the chars for indentation.

vahidR commented:
aha.. that's why then. I'll fix it

vahidR commented:
All done... I hope it's the last update to PR :) 
I corrected my IDE to use tabs rather than spaces..

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,78 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+		List<String> matchedHeaders = new ArrayList<>();
+		String headers = msg.getResponseHeader().toString();
+		String[] headerElements = headers.split("\\r\\n");
+		Pattern pattern = Pattern.compile("X-Powered-By.*", Pattern.CASE_INSENSITIVE);
```
The regex could enforce that the line starts with the header `^X-Powered-By.*` to include always the correct values:
```
HTTP/1.1 200 OK
Server: X-Powered-By/1.1 ;)
X-Powered-By: PHP/5.4
```
But see next comment.

vahidR replied:
Agreed and done..

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,78 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+		String headers = msg.getResponseHeader().toString();
+		String[] headerElements = headers.split("\\r\\n");
+		Pattern pattern = Pattern.compile("X-Powered-By.*", Pattern.CASE_INSENSITIVE);
+		Matcher matcher = pattern.matcher(headers);
```
Shouldn't this statement be moved to inside the for loop and use `header` instead of `headers`? Otherwise the use of `headerElements` is not really needed (one could keep `find()`ing the next match).

vahidR replied:
Refactored..

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java:
```diff
@@ -58,30 +61,78 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+	/**
+	 * Raises an alert with the "Evidence" or "Other" field filled-in depending on the header repetition.
+	 * @param msg The Http message containing the response headers
+	 * @param id The alert ID
```
This is the ID of the message being scanned.

vahidR replied:
Corrected..

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScannerUnitTest.java:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	@Test
+	public void shouldBeCaseSensitiveToHeaders() throws Exception {
```
`shouldBeCaseSensitiveWhenShowingHeadersInEvidenceAndOtherInfo`? (Just to be explicit that's not the header matching in the response header.)
The first header name (evidence) could also have some case changes.

vahidR replied:
Agreed and refactored the method name..

vahidR commented:
I did some refactoring that were mentioned

thc202 commented:
Looks good to me, thank you!

vahidR commented:
Thank you guys for the careful reviewing... 
I will follow your points in the future PRs

thc202 commented:
Just one thing that I missed, the ZapAddOn.xml file needs to be updated (to mention the change in the scanner and the version bumped).

vahidR commented:
> Just one thing that I missed, the ZapAddOn.xml file needs to be updated (to mention the change in the scanner and the version bumped).


OK.. Let me check how it should be done and update the PR

vahidR commented:
As there were no `issue` assigned for this changes, should I mention the PR number (which is #483) in the body of `<changes>`

thc202 commented:
Just the description is fine.

vahidR commented:
The `ZapAddOn.xml` file updated to reflect the changes..

kingthorin commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/ZapAddOn.xml:
```diff
@@ -9,6 +9,8 @@
 	<![CDATA[
 		Issue 2538: Strict-Transport-Security Header Scanner additional variants.<br>
 		Issue 2716: Implement a passive check for cookies without SameSite set.<br>
```
11 was in fact released (https://github.com/zaproxy/zap-extensions/releases) so the previous change entries can be removed.

vahidR replied:
@kingthorin Just to confirm, by "previous change entries" you mean the line 10 and 11, and not only 11, right ?

thc202 replied:
Yes, both lines.
Btw, the "changes" entries should be short descriptions (the UI does not word wrap that).

vahidR replied:
Alright ... I will make it short and remove those lines

kingthorin replied:
Yup sorry for the confusion, when I mentioned 11 I meant v11 of the addon. So yes both lines 10 and 11 can be removed.

<sub> (https://github.com/zaproxy/zap-extensions/blob/3f9572289db0eb3a78a6d7557a73ddeb2f786d7e/src/org/zaproxy/zap/extension/pscanrulesAlpha/ZapAddOn.xml#L10-L11)</sub>

kingthorin commented:
Yup sorry for the confusion, when I mentioned 11 I meant v11 of the addon. So yes both lines 10 and 11 can be removed.

<sub> (https://github.com/zaproxy/zap-extensions/blob/3f9572289db0eb3a78a6d7557a73ddeb2f786d7e/src/org/zaproxy/zap/extension/pscanrulesAlpha/ZapAddOn.xml#L10-L11)</sub>

vahidR commented:
I changed the ZapAddOn.xml file. There is nothing left :)

thc202 commented:
Looks good to me, thanks!
